File: lib/anchor.mjs
import { spawnSync } from 'node:child_process';
import path from 'node:path';
import { existsSync } from 'node:fs';
import { appendRecord, chainHead, CHAIN_FILE } from './chain.mjs';

const GIT_CANDIDATES = [
  'git',
  'C:\\Program Files\\Git\\cmd\\git.exe',
  'C:\\Program Files (x86)\\Git\\cmd\\git.exe',
];

export function findGit() {
  for (const candidate of GIT_CANDIDATES) {
    try {
      const result = spawnSync(candidate, ['--version'], { encoding: 'utf8' });
      if (result.status === 0) return candidate;
    } catch {
      // try next candidate
    }
  }
  return null;
}

export async function runAnchor(cwd, { message = null } = {}) {
  if (!existsSync(path.join(cwd, '.git'))) {
    throw new Error('not a git repository — run `git init` first');
  }
  const git = findGit();
  if (!git) throw new Error('git executable not found');

  const head = await chainHead(cwd);
  const record = await appendRecord(cwd, 'anchor', { head, message });

  const subject = `poc-evidence anchor ${head.slice(0, 16)}${message ? `: ${message}` : ''}`;
  const run = (args) => spawnSync(git, args, { cwd, encoding: 'utf8' });

  let result = run(['add', '-f', CHAIN_FILE]);
  if (result.status !== 0) {
    throw new Error(`git add failed: ${result.stderr || result.error?.message || 'unknown error'}`);
  }
  result = run(['commit', '--allow-empty', '-m', subject]);
  if (result.status !== 0) {
    throw new Error(`git commit failed: ${result.stderr || result.error?.message || 'unknown error'}`);
  }
  const commit = run(['rev-parse', 'HEAD']).stdout.trim();

  return { record, commit, head, subject };
}

File: cpp/anchor.cpp
// anchor.cpp — see anchor.h.
#include "anchor.h"

#include "chain.h"
#include "jsjson.h"
#include "util.h"

#include <stdexcept>

namespace pocev {

namespace {

std::string findGit() {
    static const char* candidates[] = {
        "git",
        "C:\\Program Files\\Git\\cmd\\git.exe",
        "C:\\Program Files (x86)\\Git\\cmd\\git.exe",
    };
    for (const char* candidate : candidates) {
        try {
            CommandResult r = runCommand({candidate, "--version"}, "");
            if (!r.signaled && r.code == 0) return candidate;
        } catch (const std::runtime_error&) {
            // try next candidate
        }
    }
    return std::string();
}

}  // namespace

AnchorResult runAnchor(const std::string& cwd, const std::string& message, bool hasMessage) {
    if (!pathExists(pathJoin(cwd, ".git"))) {
        throw std::runtime_error("not a git repository — run `git init` first");
    }
    std::string git = findGit();
    if (git.empty()) throw std::runtime_error("git executable not found");

    std::string head = chainHead(cwd);
    Json payload = Json::object();
    payload.set("head", Json::string(head));
    payload.set("message", hasMessage ? Json::string(message) : Json::null());
    Json record = appendRecord(cwd, "anchor", payload);

    std::string subject = "poc-evidence anchor " + head.substr(0, 16);
    if (hasMessage) subject += ": " + message;

    CommandResult result = runCommand({git, "add", "-f", CHAIN_FILE}, cwd);
    if (result.signaled || result.code != 0) {
        std::string why = result.err.empty() ? "unknown error" : result.err;
        throw std::runtime_error("git add failed: " + why);
    }
    result = runCommand({git, "commit", "--allow-empty", "-m", subject}, cwd);
    if (result.signaled || result.code != 0) {
        std::string why = result.err.empty() ? "unknown error" : result.err;
        throw std::runtime_error("git commit failed: " + why);
    }
    result = runCommand({git, "rev-parse", "HEAD"}, cwd);
    std::string commit;
    size_t start = 0;
    while (start < result.out.size()) {
        size_t nl = result.out.find('\n', start);
        std::string line = result.out.substr(
            start, nl == std::string::npos ? std::string::npos : nl - start);
        size_t s = line.find_first_not_of(" \t\r\n");
        size_t e = line.find_last_not_of(" \t\r\n");
        if (s != std::string::npos) {
            commit = line.substr(s, e - s + 1);
            break;
        }
        if (nl == std::string::npos) break;
        start = nl + 1;
    }

    AnchorResult out;
    out.record = record;
    out.commit = commit;
    out.head = head;
    out.subject = subject;
    return out;
}

}  // namespace pocev
